nano/nano_node/daemon.cpp

```diff
@@ -146,16 +144,27 @@ void nano::daemon::run (std::filesystem::path const & data_path, nano::node_flag
 				logger.info (nano::log::type::daemon, "Database backend: {}", node->store.vendor_get ());
 				logger.info (nano::log::type::daemon, "Start time: {:%c} UTC", fmt::gmtime (dateTime));
 
+				// IO context runner should be started first and stopped last to allow asio handlers to execute during node start/stop
+				runner = std::make_unique<nano::thread_runner> (io_ctx, node->config.io_threads);
+
 				node->start ();
 
-				nano::ipc::ipc_server ipc_server (*node, config.rpc);
+				std::latch latch{ 1 };
+
+				std::unique_ptr<nano::ipc::ipc_server> ipc_server = std::make_unique<nano::ipc::ipc_server> (*node, config.rpc);
 				std::unique_ptr<boost::process::child> rpc_process;
-				std::shared_ptr<nano::rpc> rpc;
 				std::unique_ptr<nano::rpc_handler_interface> rpc_handler;
+				std::shared_ptr<nano::rpc> rpc;
+
 				if (config.rpc_enable)
 				{
 					if (!config.rpc.child_process.enable)
 					{
+						auto stop_callback = [this, &latch] () {
+							logger.warn (nano::log::type::daemon, "RPC stop request received, stopping...");
+							latch.count_down ();
```

Comment:
count_down has undefined behavior if the amount counted down is greater than the internal counter. I think this could be a problem if RPC stop is called multiple times or is called at the same time as the signal handler.

Looking around maybe std::atomic<bool>::wait would fit better? https://en.cppreference.com/w/cpp/atomic/atomic/wait

Reply:
I didn't realize that std::latch had this behavior, thanks for pointing it out. I changed it to use atomic wait.